get_smb.go

```diff
@@ -0,0 +1,340 @@
+		i := strings.LastIndex(filePath, "/")
+		file = filePath[i+1:]
+		filePath = filePath[:i]
+	} else {
+		file = filePath
+		filePath = "."
+	}
+
+	baseCmd := g.smbclientBaseCmd(u.User, hostPath, filePath)
+	// check if file exists in the smb shared folder and check the mode
+	isDir, err := g.isDirectory(baseCmd, file)
+	if err != nil {
+		return 0, err
+	}
+	if isDir {
+		return ModeDir, nil
+	}
+	return ModeFile, nil
+}
+
+func (g *SmbGetter) Get(ctx context.Context, req *Request) error {
+	if req.u.Host == "" || req.u.Path == "" {
+		return new(smbPathError)
+	}
+
+	// If dst folder doesn't exists, we need to remove the created on later in case of failures
+	dstExisted := false
+	if req.Dst != "" {
+		if _, err := os.Lstat(req.Dst); err == nil {
+			dstExisted = true
+		}
+	}
+
+	// First look in a possible local mount of the shared folder
+	path := "/" + req.u.Host + req.u.Path
+	if runtime.GOOS == "windows" {
+		path = "/" + path
+	}
+	f := new(FileGetter)
+	result := f.get(path, req)
+	if result == nil {
+		return nil
+	}
+
+	// If not mounted, try downloading the directory content using smbclient cli
+	err := g.smbclientGet(req)
+	if err == nil {
+		return nil
+	}
+
+	result = multierror.Append(result, err)
+
+	if !dstExisted {
+		// Remove the destination created for smbclient
+		os.Remove(req.Dst)
+	}
+
+	return &smbGeneralError{result}
+}
+
+func (g *SmbGetter) smbclientGet(req *Request) error {
+	hostPath, directory, err := g.findHostAndFilePath(req.u)
+	if err != nil {
+		return err
+	}
+
+	baseCmd := g.smbclientBaseCmd(req.u.User, hostPath, ".")
+	// check directory exists in the smb shared folder and is a directory
+	isDir, err := g.isDirectory(baseCmd, directory)
+	if err != nil {
+		return err
+	}
+	if !isDir {
+		return fmt.Errorf("%s source path must be a directory", directory)
+	}
+
+	// download everything that's inside the directory (files and subdirectories)
+	smbclientCmd := baseCmd + " --command 'prompt OFF;recurse ON; mget *'"
+
+	if req.Dst != "" {
+		_, err := os.Lstat(req.Dst)
+		if err != nil {
+			if os.IsNotExist(err) {
+				// Create destination folder if it doesn't exist
+				if err := os.MkdirAll(req.Dst, 0755); err != nil {
+					return fmt.Errorf("failed to create destination path: %s", err.Error())
+				}
+			} else {
+				return err
+			}
+		}
+	}
+
+	_, err = g.runSmbClientCommand(smbclientCmd, req.Dst)
+	return err
+}
+
+func (g *SmbGetter) GetFile(ctx context.Context, req *Request) error {
+	if req.u.Host == "" || req.u.Path == "" {
+		return new(smbPathError)
+	}
+
+	// If dst folder doesn't exist, we need to remove the created one later in case of failures
+	dstExisted := false
+	if req.Dst != "" {
+		if _, err := os.Lstat(req.Dst); err == nil {
+			dstExisted = true
+		}
+	}
+
+	// First look in a possible local mount of the shared folder
+	path := "/" + req.u.Host + req.u.Path
+	if runtime.GOOS == "windows" {
+		path = "/" + path
+	}
+	f := new(FileGetter)
+	result := f.getFile(path, req, ctx)
```

Comment:
Since I'm using `go-multierror` I didn't want to override the previous `err` and result was the multierror variable. I created now an explicitly var to this to make it look better, but with the same name though 🤔